### PR TITLE
feat: add operation messages via `RAISE NOTICE`

### LINKED ISSUE #12: Allow messages to be registered from any SQL function
One of the great things about this plugin is that it allows pre/post mutation functions to "register" one or more notification messages that can be passed back to the GraphQL client. Crucially, these messages **augment** (not replace) the mutation response, which means that the mutation can still return what the client expects (e.g. the new entity in the case of an insert) PLUS any optional relevant notification messages.

It would be great if this convenience could be extended to ALL mutation functions. In particular, I'd like for any database function to be able to register one or more notification messages to an "ambient" transaction-specific context that could ultimately be read by PostGraphile (at the end of the request/response lifecycle, when generating the GraphQL response) to add those messages to the response, just like it currently does with the operation hooks functions.

Messages could be added via a universally accessible convenience function accepting the same four properties of the current notification type. For example, `add_notification(type, message, code, path)` would add a notification message to the "ambient" transaction-specific context.

Ideally, messages could also be checked via similar convenience functions, so that database functions may check if messages have been added by other functions. For example `get_notifications()` or `get_notification(type)` or even `has_notification_error()` could be used by a function to check if another called function added a notification of type error that should cause the transaction to abort or short-circuit.

Of course, there is the question of what an "ambient" transaction-specific context is, and how it would be implemented. By "ambient" I mean a context that can be accessed directly and from any function, that is, without having to pass it explicitly from function to function. In this sense, it would be similar to any global variable or settings, like the ones used to store and access the JWT claims. By transaction-specific, I mean that the context would apply to the currently running transaction only, with functions not having to worry about any transactional concerns.

As for how to implement it, there are several possibilities, but I'll leave that to the experts (I'm looking at you Benjie ;) 

Irrespective of how it's implemented, an universal mechanism to register notifications from any function would be invaluable for complex applications that need to perform non-trivial business logic. It would allow functions to keep their signatures intact (no need to pass notifications messages back and forth explicitly) while still having the ability to return any relevant notifications to the GraphQL client.

Here's some potential implementations @benjie / @demianuco discussed:

### Sending messages with LISTEN/NOTIFY

One possibility we've discussed is using PostgreSQL Listen/Notify, in which the add_notification() function would _notify_ in a channel previously created (and subscribed to) by PostGraphile, which would "accumulate" notifications thus and ultimately append them to the GraphQL response at the end of the response life-cycle. There are two problems with this approach, however, one being that the payload is limited (shouldn't be a problem, as notification messages should be small), and the other being that the mechanism will not survive a transaction rollback, thus making it impossible to send errors back to the client (show stopper!).

### Sending messages with RAISE NOTICE

Another possibility is to using the simpler `RAISE NOTICE USING ERRCODE` statement, which raises an indelible notice that is unaffected by rollbacks, and which PostGraphile can used to accumulate notification messages in a more prosaic way. It is a bit of a hack, in the sense that RAISE NOTICE was not meant for this, and the notification might pollute logs (though this could also be an advantage for development and debugging!). Also, it is perhaps fragile as it relies on the right level of verbosity being configured, with any changes to that verbosity breaking the functionality.

### Sending messages via unlogged table

Another possibility is to use an explicit global table (e.g. `tx_notifications`) to which notification messages would be added. It would obviously have to be transactionally scoped, with a column storing the current transaction ID (perhaps `select txid_current()`?), but with the transactional concerns encapsulated away by the convenience functions, so that client functions could add notification messages without having to know about transactions. The table would have no limitations in terms of payload, but I'm concerned about performance and contention issues, especially because this table would be highly volatile and in constant and concurrent use.  Perhaps the use of an unlogged table is appropriate in this scenario.

### PR DESCRIPTION
**Pull request by benjie**
Fixes #12 (cc @demianuco)

PostGraphile will subscribe to `NOTICE` events from PostgreSQL during mutations, and any notice with error code `OPMSG` (operation message) will be added to the mutation payload `messages` field.

### Important:

- This feature only applies to mutations
- Any PL/pgSQL function may raise these notices
- Additional information can be added to the OPMSG by passing a JSON value to the `detail` field
- `RAISE NOTICE` may have message length limits; it's your responsibility to colour inside the lines
- Clients must be configured (via `client_min_messages`) to receive NOTICE messages; this is PostgreSQL's default configuration.
- If the server is configured (via `log_min_messages`) to log NOTICEs then, well, the NOTICEs will be logged; PostgreSQL's default configuration does not log NOTICE.

### Example:

```sql
raise notice 
  '2 + 2 is %, minus 1 that''s %; quick maths.',
  (2 + 2),
  (2 + 2 - 1)
using
  errcode = 'OPMSG',
  detail = json_build_object(
    'level', 'info',
    'path', array_to_json(ARRAY['noticePath']),
    'anything_else', 'can_go_here'
  )::text;
```

### CODE REVIEW

**Comment by benjie** [comment]
Rebased on #17 (after merge to master) to make the diff smaller and more focussed.

**Comment by mathroc** [comment]
I don’t understand how notices are filtered, it looks like code different from `OPMSG` would be processed too, am I missing something ?

**Comment by benjie** [comment]
:eyes: I was so tired that day. I wasn't confident merging because I knew I'd overlooked something but couldn't figure out what. Thanks for spotting @mathroc! Added test and fixed.

File: src/PgNoticeMessagesPlugin.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+      };
+      let processNotice: ((msg: any) => void) | null = null;
+
+      const registerNotifyListener: OperationHookCallback = (
+        input,
+        _args,
+        context,
+        resolveInfo: GraphQLResolveInfoWithMessages
+      ) => {
+        assert(!processNotice, "processNotice should not be set yet!");
```
Comment (demianuco) [inline src/PgNoticeMessagesPlugin.ts]: This assertion passes on the first request, but fails on subsequent ones. Commenting it out does not seem to have any ill effects, and makes everything work as expected.

Comment (benjie) [reply src/PgNoticeMessagesPlugin.ts]: Fixed :+1:

**Pull request merged**